Scripts/Runtime/Entities/PersistentData/Data/EntityPersistentData.cs

```diff
@@ -1,39 +1,33 @@
 using Anvil.Unity.DOTS.Entities.TaskDriver;
+using Anvil.Unity.DOTS.Jobs;
 using Unity.Collections;
 using Unity.Collections.LowLevel.Unsafe;
 using Unity.Entities;
+using Unity.Jobs;
 
 namespace Anvil.Unity.DOTS.Entities
 {
     internal class EntityPersistentData<T> : AbstractTypedPersistentData<UnsafeParallelHashMap<Entity, T>>,
-                                             IEntityPersistentData<T>
-        where T : struct
+                                             IDriverEntityPersistentData<T>,
+                                             ISystemEntityPersistentData<T>,
+                                             IWorldEntityPersistentData<T>
+        where T : struct, IEntityPersistentDataInstance
     {
-        private readonly IEntityPersistentData<T>.DisposalCallbackPerEntity m_DisposalCallbackPerEntity;
-
-        public EntityPersistentData(string id, IEntityPersistentData<T>.DisposalCallbackPerEntity disposalCallbackPerEntity) : base(id, new UnsafeParallelHashMap<Entity, T>(ChunkUtil.MaxElementsPerChunk<Entity>(), Allocator.Persistent))
+        public EntityPersistentData()
+            : base(new UnsafeParallelHashMap<Entity, T>(ChunkUtil.MaxElementsPerChunk<Entity>(), Allocator.Persistent))
         {
-            m_DisposalCallbackPerEntity = disposalCallbackPerEntity;
         }
 
         protected override void DisposeData()
         {
-            if (m_DisposalCallbackPerEntity != null)
+            ref UnsafeParallelHashMap<Entity, T> data = ref Data;
+            foreach (KeyValue<Entity, T> entry in data)
             {
-                ref UnsafeParallelHashMap<Entity, T> data = ref Data;
-                foreach (KeyValue<Entity, T> entry in data)
-                {
-                    m_DisposalCallbackPerEntity(entry.Key, entry.Value);
-                }
+                entry.Value.DisposeForEntity(entry.Key);
```

Comment:
No boxing because we're working with the explicit `UnsafeParallelHashMap<Entity, T>` which by virtue of using this class, we know what T is. 

The boxing only happens if you do a `foreach` on an interface like `IEnumerable` since it then uses the interface function for getting the enumerator. 

Because we have the explicit collection, duck typing gets us the right Enumerator call which returns the strong typed version.

Reply:
Ok, notice that the call is different though in IL. It's a class call vs a value call

Reply:
Weird, I'm not sure to be honest. I did just check in the profiler and we're not doing any GC Allocs